src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +85,581 @@ impl ReductionPushdown {
+            self.new_inputs[idx] = inner.take_dangerous();
+        }
+        // Open question: is it useful to propagate the expected group size from
+        // the outer reduce?
+        // Is it worthwhile to propagate monotonicity from the outer reduce?
+        // If an outer reduce is not monotonic, an inner reduce can be monotonic
+        // or not monotonic, so there should be a separate pass to check
+        // monotonicity or inner reduces anyway.
+        self.new_inputs[idx] = MirRelationExpr::Reduce {
+            input: Box::new(self.new_inputs[idx].clone()),
+            group_key: inner_group_keys.clone(),
+            aggregates: inner_aggs.iter().map(|(_, agg)| agg.clone()).collect(),
+            monotonic: false,
+            expected_group_size: None,
+        };
+        // calculate new permutation. Register expressions evaluated in the
+        // inner reduce
+        for group_key in inner_group_keys {
+            if let MirScalarExpr::Column(c1) = group_key {
+                self.permutation.insert(
+                    self.old_join_mapper.map_column_to_global(c1, idx),
+                    self.next_new_join_column,
+                );
+            } else {
+                self.evaluated_expressions.insert(
+                    self.old_join_mapper.map_expr_to_global(group_key, idx),
+                    MirScalarExpr::Column(self.next_new_join_column),
+                );
+            }
+            self.next_new_join_column += 1;
+        }
+        for (agg_num, agg) in inner_aggs {
+            let agg_lookup_key = self.old_join_mapper.total_columns() + agg_num;
+            let (outer_agg_func, downcast) = agg.func.outer_agg();
+            // calculate aggs to add to outer reduce
+            let outer_agg = AggregateExpr {
+                func: outer_agg_func,
+                expr: MirScalarExpr::Column(agg_lookup_key),
+                distinct: agg.distinct,
+            };
+            if let Some(downcast) = downcast {
+                self.downcast_partial_reduce.push(MirScalarExpr::CallUnary {
+                    func: downcast,
+                    expr: Box::new(MirScalarExpr::Column(self.outer_group_key.len() + agg_num)),
+                });
+            }
+            self.outer_aggs.push((agg_num, outer_agg));
+            self.permutation
+                .insert(agg_lookup_key, self.next_new_join_column);
+            self.next_new_join_column += 1;
+        }
+        self.pushdown_succeeded = true;
+    }
+}
+
+/// Construct a graph showing, for each input, which other inputs it is joined
+/// to and on which join keys.
+/// Example:
+/// For the query
+/// """
+/// select *
+/// from foo, bar, baz
+/// where foo.a = bar.b
+/// and foo.c = bar.d
+/// and bar.e = baz.f
+/// """
+/// The corresponding graph will be
+/// "foo" -> ["bar" -> ["a", "c"]]
+/// "bar" -> ["foo" -> ["b", "d"], "baz" -> ["e"]]
+/// "baz" -> ["baz" -> ["f"]]
```

Comment:
i think this should be

`"baz" -> ["bar" -> ["f"]]`

Reply:
maybe consider a doctest here?

Reply:
Thanks for catching that!

I did consider adding a doctest, but my worry is that it becomes less clear what the graph is meant to look like once it becomes bunch of 0 -> [1 -> [`ScalarExpr::Column(0)`, `ScalarExpr::Column(1)`]].